### PR TITLE
test: Add unified spec runner scaffolding

### PR DESCRIPTION
**Pull request by nbbeeken**
Using the unified spec test schema this scaffolding
outlines the structure of a unified runner. Most tests
are skipped by the runOn requirements or not implemented
errors thrown by empty operation functions.

NODE-2287

### CODE REVIEW

**Review by durran — CHANGES_REQUESTED** [review]

File: test/functional/unified-spec-runner/entities.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    const m = new Map<string, UnifiedMongoClient>();
+    for (const [id, e] of this) {
+      if (e instanceof UnifiedMongoClient) {
+        m.set(id, e);
+      }
+    }
+    return m;
+  }
+
+  getClient(key: string): UnifiedMongoClient {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/functional/unified-spec-runner/entities.ts]: I agree, each of these methods also means we'd have a different stack trace for each one even though it is the same error, I've consolidated the helpers into `getEntity` and `mapOf` and I think it's friendly enough we don't need the bespoke methods any more, getEntity with a type param works nicely!

File: test/functional/unified-spec-runner/operations.ts [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    if (operation.expectError) {
+      expect.fail(`Operation ${operationName} succeeded but was not supposed to`);
+    }
+
+    if (operation.expectResult) {
+      if (isSpecialOperator(operation.expectResult)) {
+        specialCheck(result, operation.expectResult);
+      } else {
+        for (const [resultKey, resultValue] of Object.entries(operation.expectResult)) {
+          // each key/value expectation can be is a special op
```
Comment (emadum) [inline test/functional/unified-spec-runner/operations.ts:312]:
```suggestion
          // each key/value expectation can be a special op
```

**Review by emadum — APPROVED** [review]
Suggested a few typo fixes in comments but LGTM 👍

**Review by durran — APPROVED** [review]
Nice refactor! LGTM

**Pull request merged**